java/arcs/core/allocator/Arc.kt

```diff
@@ -53,19 +56,18 @@ class Arc internal constructor(
      *  is considered to be [Error].
      */
     var arcState: ArcState
-        get() = arcStateInternal
-        private set(state) = sync(this) {
-            if (arcStateInternal != state) {
-                arcStateInternal = state
-                fireArcStateChange()
-            }
+        get() = arcStateInternal.value
+        private set(state) {
+            arcStateInternal.update { state }
         }
 
-    private fun onArcStateChange(handler: (ArcState) -> Unit) = sync(this) {
-        if (arcStateChangeHandlers.isEmpty()) {
-            registerChangeHandlerWithArcHosts()
+    private suspend fun onArcStateChange(handler: (ArcState) -> Unit) {
+        maybeRegisterChangeHandlerWithArcHosts()
+        arcStateChangeHandlers.update {
+            val newHandlers = it.toMutableList()
+            newHandlers += handler
+            newHandlers
```

Comment:
Can't this just be `arcStateChangeHandlers.update { it + handler }`?

Reply:
Yup. I used to have this as 
val newHandlers = it.toMutableList()
 newHandlers += handler

arcStateChangeHandlers.update { newHandlers }

and then I inlined it. I was earlier avoiding doing 'too much' inside the update, but I ended up inlining it and not cleaning it up.